test/distributed/fsdp/test_fsdp_clip_grad_norm.py

```diff
@@ -39,67 +41,133 @@
+            out = model(*inp)
+            loss = model.module.get_loss(inp, out)
+            loss.backward()
 
-    @skip_if_lt_x_gpu(2)
-    @parametrize("norm_type", [2.0, inf])
-    @parametrize("nested_fsdp", [True, False])
-    @parametrize(
-        "cpu_offload",
-        [CPUOffload(offload_params=True), CPUOffload(offload_params=False)],
-    )
-    def test_fsdp_clip_grad_norm(self, norm_type, nested_fsdp, cpu_offload):
-        """Test FSDP with clip grad norm."""
-        self._run_fsdp_one_iteration(norm_type, nested_fsdp, cpu_offload)
+        # Multiply gradients by a large factor to ensure that gradients will
+        # actually be clipped
+        for param in itertools.chain(ddp_model.parameters(), fsdp_model.parameters()):
+            if param.grad is not None:  # gradients may be `None` for `use_orig_params=True`
+                param.grad *= LARGE_FACTOR
+        orig_ddp_grads = [param.grad.detach().clone() for param in ddp_model.parameters()]
+        orig_fsdp_grads = [
+            param.grad.detach().clone() if param.grad is not None else None
+            for param in fsdp_model.parameters()
+        ]
+
+        ddp_total_norm = torch.nn.utils.clip_grad_norm_(
+            ddp_model.parameters(), max_norm=max_norm, norm_type=norm_type,
+        )
+        fsdp_total_norm = fsdp_model.clip_grad_norm_(max_norm=max_norm, norm_type=norm_type)
+        self.assertEqual(ddp_total_norm, fsdp_total_norm)
 
+        # Check that the gradients were modified by `clip_grad_norm_()`
+        for param, orig_grad in zip(ddp_model.parameters(), orig_ddp_grads):
+            assert not torch.equal(param.grad, orig_grad)
+        for param, orig_grad in zip(fsdp_model.parameters(), orig_fsdp_grads):
+            if param.grad is None:
+                self.assertEqual(param.grad, orig_grad)  # `None`
+            else:
+                assert not torch.equal(param.grad, orig_grad)
 
-class TestCalcuGradNorm(FSDPTest):
-    @skip_if_lt_x_gpu(2)
-    @parametrize("norm_type", [2.0, inf, 1.3, 2.5])
-    @parametrize("nested_fsdp", [True, False])
-    def test_fsdp_calc_grad_norm(self, norm_type, nested_fsdp):
-        """Test grad norm cal API."""
-        model = FSDP(DeterministicModel(nested_fsdp))
-        input = torch.rand(15, 2, device=self.rank)
-        out = model(input)
-        out.sum().backward()
-        total_norm = _calc_grad_norm(model.params_with_grad, norm_type)
-        total_norm_expected = _collect_total_grad_norm_local(model, norm_type)
-        self.assertEqual(total_norm, total_norm_expected)
+        # Run an optimizer step to ensure gradients matched after clipping
+        ddp_optim.step()
+        fsdp_optim.step()
+        with FSDP.summon_full_params(fsdp_model):
+            for (n1, p1), (n2, p2) in zip(
+                ddp_model.module.named_parameters(),
+                fsdp_model.named_parameters(),
+            ):
+                self.assertEqual(n1, n2)
```

Comment:
nit: checking names is nice but probably beyond the scope of this unittest?

Reply:
I agree.

To be honest, I tend to do this while testing so that `n1` and `n2` are not unused variables. It is often helpful when the subsequent parameter equality check fails to see the name of the violating parameter. The issue is that if you override the error string to add the name, you lose the normal error string. I wonder if there is a way to simply _append_ a custom string to the normal error string.